feat(searchPagination): added pagination search api for tickets api

naveenreddyalka opened the pull request:
**Description**

Fixes [#DUNLOP-14219](https://team-turo.atlassian.net/browse/DUNLOP-14219)

Add pagination to search api
  Add sort and page params options
  New Page/TicketPage class to capture content, count, next page urls

Commit:
feat(searchPagination): added pagination search api fpr tickets

Commit:
feat(refactor): remove method finals and pom changes

h-wang94 commented on src/main/java/org/zendesk/client/v2/Zendesk.java:
```diff
@@ -1750,6 +1752,63 @@ public void deleteOrganizationMembership(long user_id, long organization_members
     }
 
+    public <T> Optional<T> getSearchPageResults(
+        final String query,
```
We should probably mark most of these as `@Nullable` ?

h-wang94 commented on src/main/java/org/zendesk/client/v2/Zendesk.java:
```diff
@@ -1750,6 +1752,63 @@ public void deleteOrganizationMembership(long user_id, long organization_members
+      final StringBuilder uriTemplate = new StringBuilder("/search.json{?query");
+
+      for (final String paramName : paramsMap.keySet()) {
+          uriTemplate.append(",")
```
This will cause the url to have a `,` in front like

```
/search.json{?query,<paramName>,<paramName2>}
```

You can use 
```
String.join(",", paramsMap.keySet())
```
instead
so you can get 
```
/search.json{?query<paramName>,<paramName2>}
```

Also, should the url look like that? Not sure how I verify/check

naveenreddyalka replied:
TemplateUri is an extension of UriTemplate written by cloud bee's team. 
It works fine as used above 
[see other usages here](https://github.com/turo/zendesk-java-client/blob/74455b24bd8ecc64fec1c88252b9327393cca1ec/src/main/java/org/zendesk/client/v2/Zendesk.java#L1735)

h-wang94 commented on src/main/java/org/zendesk/client/v2/model/Page.java:
```diff
@@ -0,0 +1,41 @@
+import java.io.Serializable;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public abstract class Page<T extends SearchResultEntity> implements Serializable {
```
Would probably add a javadoc of some sort to indicate that this is what allows us to support paging with zendesk

h-wang94 reviewed (commented):
I'm not sure if you want to maybe add a comment on some of the models you created to maybe indicate that this is divergent from the original zendesk-java-client.

Commit:
feat(refactor): add more java docs and refactoring page definition

romenrg commented on src/main/java/org/zendesk/client/v2/Zendesk.java:
```diff
@@ -1749,7 +1753,82 @@ public void deleteOrganizationMembership(long user_id, long organization_members
+      }
+
+      
+      final StringBuilder uriTemplate = new StringBuilder("/search.json{?query");
```
Seems the `uriTemplate` logic is shared with the `getSearchResults` method. Might consider extracting it to a private method used in both

romenrg commented on src/main/java/org/zendesk/client/v2/Zendesk.java:
```diff
@@ -1749,7 +1753,82 @@ public void deleteOrganizationMembership(long user_id, long organization_members
+
+      uriTemplate.append("}");
+      
+      String typeName = null;
```
Actually, this loop could also be extracted and shared with `getSearchResults`, since the only difference is `searchType.equals(entry.getValue())` vs `type.equals(entry.getValue())` and that var could be a parameter of the extracted private method

romenrg reviewed (commented):
Finally had a chance to review this.

If I understood correctly, what you are doing here is to allow pagination by keeping the pagination logic out of the library; delegating its control to the app that passes the query to the library (which should keep track of the "pointers" from the results and make as many call as necessary to `getSearchTicketPageResults` with the corresponding pagination parameters in the query passed). Did I understand that right? Seems like a valid enhancement to me, to provide an alternative to the existing implementation which seemed to always retrieve all the results (not letting the user handle the pagination).

I noticed most of the logic in `getSearchPageResults` is shared with the pre-existing `getSearchResults` (the `uriTemplate` entirely and the `typeName` loop if parametrized as explained in the comment I made in the corresponding line)

Commit:
refactor(Zendesk): updated for review comments

Commit:
refactor(Zendesk): remove usage of internal NotNull/Nullable and rename methods

aheritier committed:
Add @PierreBtz as code owner

aheritier committed:
[fix] RealSmokeTest#createOrganization test was failing

Because it was too long to review in 30sec all organizations to delete the ones with a given externalId

We now directly search for these specific orgs to delete them

aheritier committed:
Merge pull request #398 from cloudbees-oss/fix-test

[fix] RealSmokeTest#createOrganization test was failing

aheritier committed:
[fix] Do not execute integration tests when not setup

aheritier committed:
Add codeql analysis

aheritier committed:
Merge pull request #399 from aheritier/fix-it

[fix] Do not execute integration tests when not setup

aheritier committed:
[chore] Split PRs and Masters CI builds

aheritier committed:
[chore] Update dependabot settings

* Execute daily
* Check for github action updates

aheritier committed:
Merge pull request #400 from aheritier/fix-ci

Split PRs and Masters CI builds

dependabot[bot] committed:
Bump async-http-client from 2.12.2 to 2.12.3

Bumps [async-http-client](https://github.com/AsyncHttpClient/async-http-client) from 2.12.2 to 2.12.3.
- [Release notes](https://github.com/AsyncHttpClient/async-http-client/releases)
- [Changelog](https://github.com/AsyncHttpClient/async-http-client/blob/master/CHANGES.md)
- [Commits](https://github.com/AsyncHttpClient/async-http-client/compare/async-http-client-project-2.12.2...async-http-client-project-2.12.3)

Signed-off-by: dependabot[bot] <support@github.com>

aheritier committed:
Merge pull request #401 from aheritier/dependabot

Update dependabot settings

dependabot[bot] committed:
Bump release-drafter/release-drafter from 5.11.0 to 5.15.0

Bumps [release-drafter/release-drafter](https://github.com/release-drafter/release-drafter) from 5.11.0 to 5.15.0.
- [Release notes](https://github.com/release-drafter/release-drafter/releases)
- [Commits](https://github.com/release-drafter/release-drafter/compare/v5.11.0...v5.15.0)

---
updated-dependencies:
- dependency-name: release-drafter/release-drafter
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump actions/setup-java from 1 to 2.3.1

Bumps [actions/setup-java](https://github.com/actions/setup-java) from 1 to 2.3.1.
- [Release notes](https://github.com/actions/setup-java/releases)
- [Commits](https://github.com/actions/setup-java/compare/v1...v2.3.1)

---
updated-dependencies:
- dependency-name: actions/setup-java
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump awaitility from 4.0.3 to 4.1.0

Bumps [awaitility](https://github.com/awaitility/awaitility) from 4.0.3 to 4.1.0.
- [Release notes](https://github.com/awaitility/awaitility/releases)
- [Changelog](https://github.com/awaitility/awaitility/blob/master/changelog.txt)
- [Commits](https://github.com/awaitility/awaitility/commits/awaitility-4.1.0)

Signed-off-by: dependabot[bot] <support@github.com>

aheritier committed:
Add distribution for setup-java v2

aheritier committed:
Merge pull request #402 from cloudbees-oss/dependabot/github_actions/release-drafter/release-drafter-5.15.0

Bump release-drafter/release-drafter from 5.11.0 to 5.15.0

aheritier committed:
Java 8 instead of 1.8

aheritier committed:
Merge pull request #382 from cloudbees-oss/dependabot/maven/org.asynchttpclient-async-http-client-2.12.3

Bump async-http-client from 2.12.2 to 2.12.3

aheritier committed:
Merge pull request #385 from cloudbees-oss/dependabot/maven/org.awaitility-awaitility-4.1.0

Bump awaitility from 4.0.3 to 4.1.0

dependabot[bot] committed:
Bump jacoco-maven-plugin from 0.8.6 to 0.8.7

Bumps [jacoco-maven-plugin](https://github.com/jacoco/jacoco) from 0.8.6 to 0.8.7.
- [Release notes](https://github.com/jacoco/jacoco/releases)
- [Commits](https://github.com/jacoco/jacoco/compare/v0.8.6...v0.8.7)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump assertj-core from 3.19.0 to 3.21.0

Bumps [assertj-core](https://github.com/assertj/assertj-core) from 3.19.0 to 3.21.0.
- [Release notes](https://github.com/assertj/assertj-core/releases)
- [Commits](https://github.com/assertj/assertj-core/compare/assertj-core-3.19.0...assertj-core-3.21.0)

---
updated-dependencies:
- dependency-name: org.assertj:assertj-core
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

aheritier committed:
Merge pull request #403 from cloudbees-oss/dependabot/github_actions/actions/setup-java-2.3.1

Bump actions/setup-java from 1 to 2.3.1

dependabot[bot] committed:
Bump jackson-bom from 2.12.2 to 2.13.0

Bumps [jackson-bom](https://github.com/FasterXML/jackson-bom) from 2.12.2 to 2.13.0.
- [Release notes](https://github.com/FasterXML/jackson-bom/releases)
- [Commits](https://github.com/FasterXML/jackson-bom/compare/jackson-bom-2.12.2...jackson-bom-2.13.0)

---
updated-dependencies:
- dependency-name: com.fasterxml.jackson:jackson-bom
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

aheritier committed:
Merge pull request #386 from cloudbees-oss/dependabot/maven/org.jacoco-jacoco-maven-plugin-0.8.7

Bump jacoco-maven-plugin from 0.8.6 to 0.8.7

aheritier committed:
Merge pull request #395 from cloudbees-oss/dependabot/maven/org.assertj-assertj-core-3.21.0

Bump assertj-core from 3.19.0 to 3.21.0

aheritier committed:
Merge pull request #397 from cloudbees-oss/dependabot/maven/com.fasterxml.jackson-jackson-bom-2.13.0

Bump jackson-bom from 2.12.2 to 2.13.0

dependabot[bot] committed:
Bump slf4j-api from 1.7.30 to 1.7.32

Bumps [slf4j-api](https://github.com/qos-ch/slf4j) from 1.7.30 to 1.7.32.
- [Release notes](https://github.com/qos-ch/slf4j/releases)
- [Commits](https://github.com/qos-ch/slf4j/compare/v_1.7.30...v_1.7.32)

---
updated-dependencies:
- dependency-name: org.slf4j:slf4j-api
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

aheritier committed:
Merge pull request #404 from cloudbees-oss/dependabot/maven/org.slf4j-slf4j-api-1.7.32

Bump slf4j-api from 1.7.30 to 1.7.32
